### PR TITLE
Unflatten implicitly in the `zeek-tsv` parser

### PR DESCRIPTION
**Pull request by dominiklohmann**
Fixes tenzir/issues#615

### CODE REVIEW

File: libtenzir/builtins/formats/zeek_tsv.cpp [hunk 1]
```diff
@@ -648,12 +647,7 @@ auto parser_impl(generator<std::optional<std::string_view>> lines,
           }
           return false;
         });
-      if (target_schema != ctrl.schemas().end()
-          and can_cast(document.builder->schema(), *target_schema)) {
-        document.target_schema = *target_schema;
-      } else {
-        document.target_schema = document.builder->schema();
-      }
+      document.target_schema = *target_schema;
```
Comment (jachris) [inline libtenzir/builtins/formats/zeek_tsv.cpp:650]: Possibly related to this change, which I do not understand?

**Review by jachris — CHANGES_REQUESTED** [review]
When I tested this, I got a segfault.

~~~
#separator \x09
#set_separator	,
#empty_field	(empty)
#unset_field	-
#path	tenzir.json
#open	2023-10-10-12-48-16.056854
#fields	a.b	a.b_1
#types	int	int
42	43
#close	2023-10-10-12-48-16.057063
~~~

~~~
$ build/release/bin/tenzir --console-verbosity=VERBOSE '--console-format=%^[%s:%#] %v%$' -q 'from file zeek.tsv read zeek-tsv'
tenzir-v4.2.0-184-ge9febb8018-dirty: Error: signal 11 (Segmentation fault: 11)
0xc248800106890d00: (fatal_handler+0x50)
0xd0168001a040ea24: (_sigtramp+0x38)
0x24010001066d338c: (bool tenzir::operator==<tenzir::record_type, tenzir::record_type>(tenzir::record_type const&, tenzir::record_type const&)+0x6c)
0x71060001066d338c: (bool tenzir::operator==<tenzir::record_type, tenzir::record_type>(tenzir::record_type const&, tenzir::record_type const&)+0x6c)
0x8e0d8001066d3024: (tenzir::cast(tenzir::table_slice, tenzir::type const&)+0xbc)
0xda7c0001041ba064: (tenzir::plugins::zeek_tsv::(anonymous namespace)::parser_impl(tenzir::generator<std::__1::optional<std::__1::basic_string_view<char, std::__1::char_traits<char>>>>, tenzir::operator_control_plane&) (.resume)+0x1150)
0x8777000106ab5638: (caf::detail::default_action_impl<tenzir::(anonymous namespace)::exec_node_state<caf::intrusive_ptr<tenzir::chunk>, tenzir::table_slice>::schedule_run()::'lambda'()>::run()+0x534)
~~~

File: libtenzir/builtins/formats/zeek_tsv.cpp (hunk 1 above)
Comment (dominiklohmann) [reply libtenzir/builtins/formats/zeek_tsv.cpp]: Looks like I pushed some in-between state, sorry 🙈 I pushed everything now. Mind taking another look?

Comment (jachris) [reply libtenzir/builtins/formats/zeek_tsv.cpp]: This works now, thanks :+1:

**Review by jachris — APPROVED** [review]
Another stacktrace:

~~~
#separator \x09
#set_separator	,
#empty_field	(empty)
#unset_field	-
#path	tenzir.json
#open	2023-10-10-12-48-16.056854
#fields	a.b	a.b
#types	int	int
42	43
#close	2023-10-10-12-48-16.057063
~~~

~~~
/tmp/apache-arrow-20230828-19076-j90x0g/apache-arrow-13.0.0/cpp/src/arrow/result.cc:28: ValueOrDie called on an error: Invalid: Can't infer struct array length with 0 child arrays
0   libarrow.1300.0.0.dylib             0x000000010f6b2774 _ZN5arrow4util7CerrLog14PrintBackTraceEv + 44
1   libarrow.1300.0.0.dylib             0x000000010f6b2724 _ZN5arrow4util7CerrLogD2Ev + 184
2   libarrow.1300.0.0.dylib             0x000000010f6b2664 _ZN5arrow4util7CerrLogD0Ev + 12
3   libarrow.1300.0.0.dylib             0x000000010f6b2600 _ZN5arrow4util8ArrowLogD1Ev + 48
4   libarrow.1300.0.0.dylib             0x000000010f5eb034 _ZN5arrow8internal14DieWithMessageERKNSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEE + 64
5   libarrow.1300.0.0.dylib             0x000000010f5eb0cc _ZN5arrow8internal17InvalidValueOrDieERKNS_6StatusE + 116
6   tenzir                              0x0000000101dce05c _ZNO5arrow6ResultINSt3__110shared_ptrINS_11StructArrayEEEE10ValueOrDieEv + 52
7   libtenzir.2818.0.dylib              0x0000000124a6c950 _ZNK6tenzir12_GLOBAL__N_115unflatten_field8to_arrowEv + 1440
8   libtenzir.2818.0.dylib              0x0000000124a69324 _ZN6tenzir22unflatten_struct_arrayENSt3__110shared_ptrIN5arrow11StructArrayEEENS0_17basic_string_viewIcNS0_11char_traitsIcEEEE + 3832
9   libtenzir.2818.0.dylib              0x0000000124a71860 _ZN6tenzir9unflattenERKNS_11table_sliceENSt3__117basic_string_viewIcNS3_11char_traitsIcEEEE + 956
10  tenzir                              0x00000001021573e8 _ZZN6tenzir7plugins8zeek_tsv12_GLOBAL__N_111parser_implENS_9generatorINSt3__18optionalINS4_17basic_string_viewIcNS4_11char_traitsIcEEEEEEEERNS_22operator_control_planeEENK4$_15clEv + 596
~~~

But I don't think this is related to this PR, but is rather an overlooked edge case in `unflatten` itself. Perhaps you can sneak in a fix (+ an integration test) for this into this PR? But otherwise, it seems good to go :+1:

**Comment by dominiklohmann** [comment]
> But I don't think this is related to this PR, but is rather an overlooked edge case in `unflatten` itself. Perhaps you can sneak in a fix (+ an integration test) for this into this PR? But otherwise, it seems good to go 👍

I looked at this for a few minutes, and as I understand it this is a bug in the zeek-tsv parser, which should not allow duplicate field names. It's just that `unflatten` is the first piece of code where a record batch with duplicate field names is problematic, but it much more generally shouldn't exist.

I think the easiest way to fix this is to switch the parser to the new builder, which can handle this transparently by overwriting fields with duplicate names. But that's a lot more work, I think, so I'll write a follow-up issue for it.

**Comment by jachris** [comment]
Oh yea, right... Delegating this to a follow-up issue sounds good.
